### PR TITLE
attic-client: init at unstable-2024-02-08

### PR DESCRIPTION
**Pull request by aciceri**
## Description of changes

This adds a derivation for [https://github.com/zhaofengli/attic](https://github.com/zhaofengli/attic) that produces the following binaries:
- `attic`
- `atticdm`
- `atticd`

## Things done

Not tested yet (only `attic --version`).

`meta.mainProgram` is set to `attic` (the client) since it's probably what most people expect to find building `attic`.
To be honest this is the first time I package something in Rust that produces multiple binaries and I don't know if it makes more sense keeping all the them in the same derivation or splitting it.
If there are systems where one of the binaries doesn't build/work it could make sense.

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

@zhaofengli I hope you will appreciate this PR. Are you interested into being a nixpkgs maintainer for this?

### CODE REVIEW

**Comment by zhaofengli** [comment]
Hi, thanks for the PR! I do wish that this is hold off before the first stable release where there could be some breaking changes. I've been putting this off for a while but hopefully I'll have more time this holiday season.

**Comment by aciceri** [comment]
Hi, thank you for the useful application!
Anyway the reason why we want it in `nipxkgs` is [this PR](https://github.com/hercules-ci/hercules-ci-effects/pull/163) where we don't want to add `attic` as flake input.
But if it's a just a matter of waiting this holiday season then no problem, if it will take months instead I would prefer that this was merged as it is. I promise I will help (as far as I can with Rust) with another PR when the stable version will be released.

**Comment by aciceri** [comment]
@zhaofengli are there news on this?

**Comment by niklaskorz** [comment]
> Hi, thanks for the PR! I do wish that this is hold off before the first stable release where there could be some breaking changes. I've been putting this off for a while but hopefully I'll have more time this holiday season.

In my opinion that is not a blocker, the package is marked as unstable so it's already explicit there will be breaking changes.

**Comment by aciceri** [comment]
If you agree I can update the package to the latest version and make it ready for reviews.

**Comment by Hayajiro** [comment]
I'd love to see this packaged, as I'd want to push stuff to our cache using GitHub Actions, but the current way of blindly pulling a specific derivation and using that doesn't feel right, at all. Even adding it as a flake input (and trusting the "staging" Attic Instance as a substituter) results in us having to build attic-client ourselves.

Given that it's explicitly marked as unstable, I also don't really see a reason to further delay packaging. On the contrary, it'd make using attic just so much easier.

In case we really want to delay this further, I'll probably need to push this to my own cache and then fetch binaries from there... Not ideal, but the only alternative.

**Comment by aciceri** [comment]
Apparently this
```nix
lockFile = "${src}/Cargo.lock";
```
was causing an IFD to occur (probably the lockfile is parsed by nix at eval time?) and bot `nixpkgs-rev` and the CI were failing. Now I copied the lockfile PR's branch and did this
```nix
lockFile = ./Cargo.lock;
```
The nixpkgs manual suggests it (and I found many other packages doing the same).

I also tried adding `passthru.updateScript = nix-update-script { };` and manually running it probably expects that the upstream GitHub repository has releases or tags, so it failed and I haven't added it.

```
~/projects/aciceri/nixpkgs:attic? λ nix run n#nixpkgs-review -- pr 274481
$ git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/274481/head:refs/nixpkgs-review/1
$ git worktree add /home/ccr/.cache/nixpkgs-review/pr-274481-3/nixpkgs 3f86dedcc507143018a900cee727dc64e15df30d
Preparing worktree (detached HEAD 3f86dedcc507)
Updating files: 100% (40430/40430), done.
HEAD is now at 3f86dedcc507 Merge pull request #295182 from r-ryantm/auto-update/earthly
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f <nixpkgs> --nix-path nixpkgs=/home/ccr/.cache/nixpkgs-review/pr-274481-3/nixpkgs nixpkgs-overlays=/tmp/tmpfpsdvutc -qaP --xml --out-path --show-trace --no-allow-import-from-derivation
$ git merge --no-commit --no-ff 71fac07422fdfe21aa6567680e9e0cfdf13790d2
Automatic merge went well; stopped before committing as requested
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f <nixpkgs> --nix-path nixpkgs=/home/ccr/.cache/nixpkgs-review/pr-274481-3/nixpkgs nixpkgs-overlays=/tmp/tmpfpsdvutc -qaP --xml --out-path --show-trace --no-allow-import-from-derivation --meta
1 package added:
attic (init at 2024-2-8)

$ nix build --nix-path nixpkgs=/home/ccr/.cache/nixpkgs-review/pr-274481-3/nixpkgs nixpkgs-overlays=/tmp/tmpfpsdvutc --extra-experimental-features nix-command no-url-literals --no-link --keep-going --no-allow-import-from-derivation --option build-use-sandbox relaxed -f /home/ccr/.cache/nixpkgs-review/pr-274481-3/build.nix

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/274481

1 package built:
attic

$ /nix/store/6r0bm8shswm9v08kzvq24ib48mx3kxmn-nix-2.18.1/bin/nix-shell /home/ccr/.cache/nixpkgs-review/pr-274481-3/shell.nix --nix-path nixpkgs=/home/ccr/.cache/nixpkgs-review/pr-274481-3/nixpkgs nixpkgs-overlays=/tmp/tmpfpsdvutc
```

**Comment by Hayajiro** [comment]
@aciceri I think it would be good to package the tools separately. You can find the individual Cargo.toml files for the respective CLI tools in the folders, for example "client" for `attic` and "server" for `atticd`. Maybe you can also have a look at the [package.nix inside the attic repo](https://github.com/zhaofengli/attic/blob/main/package.nix

If you want, I can have a look at splitting up the packages, as well. But that'll probably take at least until the weekend.

**Comment by aciceri** [comment]
@Hayajiro I believe I haven't checked the most obvious place (upstream `package.nix`) before opening this PR, thanks!
So now it should support `darwin` and I splitted it into two derivations (`atticd` and `atticadm` are together since they are the same crate), I hope that overriding `attic-client` as I did is a viable approach with the new `by-name` nixpkgs structure, I took a look at the manual and at the RFC but I didn't find anything about what to do in these cases.

Also I updated `version` to `0.1.0` as in the upstream derivation, even though there are no releases or tags in the upstream repository, so I'm still using a commit hash as `fetchFromGitHub`'s `rev`.

**Comment by zhaofengli** [comment]
Please add me to the maintainer list, thanks!

**Review by Hayajiro — APPROVED** [review]
Changes LGTM so far. Will test the build output in a day or two (once I have the time to spare). That is, unless someone beats me to it.

**Review by aprilthepink — APPROVED** [review]
proud to say: just tested the client. works flawlessly

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
@ofborg build attic-client attic-server

**Comment by anthonyroussel** [comment]
Hello!
Unfortunately, looks like my PR is failing (ofborg-eval check) because of an error on the attic-client package:

```log
       error: attribute 'cxxabi' missing

       at /ofborg/checkout/1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-4/pkgs/by-name/at/attic-client/package.nix:47:24:

           46|   env = lib.optionalAttrs stdenv.cc.isClang {
           47|     NIX_LDFLAGS = "-l${stdenv.cc.libcxx.cxxabi.libName}";
             |                        ^
           48|   };
```

https://gist.github.com/GrahamcOfBorg/71643bc5b784eb2a999dd94c7e2209fc

File: pkgs/by-name/at/attic-client/package.nix
Comment (anthonyroussel) [inline pkgs/by-name/at/attic-client/package.nix:49]:
```suggestion
```

You don't need this anymore with this PR being merged: https://github.com/NixOS/nixpkgs/pull/292043

Comment (drupol) [reply pkgs/by-name/at/attic-client/package.nix:49]: Going to provide the PR fixing this.

Comment (drupol) [reply pkgs/by-name/at/attic-client/package.nix:49]: Please, provide some feedback at https://github.com/NixOS/nixpkgs/pull/299288

Comment (anthonyroussel) [reply pkgs/by-name/at/attic-client/package.nix:49]: Thanks @drupol 
I am trying to build the fix PR on aarch64-linux, then I approve it :)

**Comment by misuzu** [comment]
attic-server isn't being built by hydra for aarch64-linux for some reason: https://hydra.nixos.org/eval/1805789?filter=attic-server&compare=1805780&full=
attic-client is fine

**Pull request merged**
